libraries/net/include/psibase/shortest_path_routing.hpp

```diff
@@ -0,0 +1,676 @@
+         while (iter != end)
+         {
+            if (iter->first.peer == peer)
+            {
+               iter = routeTable.erase(iter);
+            }
+            else
+            {
+               ++iter;
+            }
+         }
+         neighborTable.erase(peer);
+         for (auto iter = selectedRoutes.begin(), end = selectedRoutes.end(); iter != end;)
+         {
+            auto next = iter;
+            ++next;
+            if (iter->second == peer)
+            {
+               selectRoute(iter->first, iter->second);
+            }
+            iter = next;
+         }
+      }
+      void send_routes(peer_id peer)
+      {
+         for (const auto& [producer, nextPeer] : selectedRoutes)
+         {
+            auto route = routeTable.find(RouteKey{producer, nextPeer});
+            if (route != routeTable.end())
+            {
+               async_send(peer,
+                          RouteUpdateMessage{producer, route->second.seqno, getMetric(*route)});
+            }
+         }
+         if (auto self = consensus().producer_name(); consensus().is_producer(self))
+         {
+            async_send(peer, RouteUpdateMessage{self, seqno, RouteMetric{0}});
+         }
+      }
+      void on_producer_change()
+      {
+         // purge out-dated routes
+         auto self = consensus().producer_name();
+         for (auto iter = selectedRoutes.begin(), end = selectedRoutes.end(); iter != end;)
+         {
+            if (consensus().is_producer(iter->first) && iter->first != self)
+            {
+               ++iter;
+            }
+            else
+            {
+               PSIBASE_LOG(logger, info) << "Removed route to " << iter->first.str();
+               iter = selectedRoutes.erase(iter);
+            }
+         }
+         for (auto iter = routeTable.begin(), end = routeTable.end(); iter != end;)
+         {
+            auto next = routeTable.upper_bound(iter->first.producer);
+            if (!consensus().is_producer(iter->first.producer) || iter->first.producer == self)
+            {
+               routeTable.erase(iter, next);
+            }
+            iter = next;
+         }
+         for (auto iter = sourceTable.begin(), end = sourceTable.end(); iter != end;)
+         {
+            if (consensus().is_producer(iter->first) && iter->first != self)
+            {
+               ++iter;
+            }
+            else
+            {
+               iter = sourceTable.erase(iter);
+            }
+         }
+         // If we became a producer, we need to establish a new route
+         if (consensus().is_producer(self))
+         {
+            multicast(RouteUpdateMessage{self, seqno, RouteMetric{0}});
+         }
+         // Sync our routing table with all peers
+         multicast(RequestRoutesMessage{});
+      }
+      bool isFeasible(std::pair<const RouteKey, RouteData>& route)
+      {
+         if (route.second.metric == RouteMetric::infinite)
+         {
+            return true;
```

Comment:
Shouldn't infinite metric return !feasible?

Reply:
https://datatracker.ietf.org/doc/html/rfc8966#feasibility-condition
"Note further that retractions (updates with infinite metric) are always feasible, since they cannot possibly cause a routing loop"